neuralmonkey/attention/scaled_dot_product.py

```diff
@@ -0,0 +1,176 @@
+        self.attention_values = get_attention_states(values_encoder)
+        self.attention_mask = get_attention_mask(keys_encoder)
+
+        self._dimension = self.attention_keys.get_shape()[-1].value
+
+        if self._dimension % self.n_heads != 0:
+            raise ValueError("Model dimension ({}) must be divisible by the "
+                             "number of attention heads ({})"
+                             .format(self._dimension, self.n_heads))
+
+        self._head_dim = int(self._dimension / self.n_heads)
+        self._scaling_factor = 1 / math.sqrt(self._head_dim)
+
+    def attention(self,
+                  query: tf.Tensor,
+                  decoder_prev_state: tf.Tensor,
+                  decoder_input: tf.Tensor,
+                  loop_state: MultiHeadLoopStateTA,
+                  step: tf.Tensor) -> Tuple[tf.Tensor, MultiHeadLoopStateTA]:
+
+        if self.n_heads == 1:
+            context, weights = self.attention_single_head(
+                query, self.attention_keys, self.attention_values)
+            head_weights = [weights]
+        else:
+            # project query, keys and vals: [batch, rnn] to [batch, rnn2]
+            head_contexts, head_weights = zip(*[
+                self.attention_single_head(
+                    tf.layers.dense(query, self._head_dim,
+                                    name="query_proj_head{}".format(i)),
+                    tf.layers.dense(self.attention_keys, self._head_dim,
+                                    name="keys_proj_head{}".format(i)),
+                    tf.layers.dense(self.attention_values, self._head_dim,
+                                    name="values_proj_head{}".format(i)))
+                for i in range(self.n_heads)])
+
+            context = tf.layers.dense(
+                tf.concat(head_contexts, -1), self._dimension,
+                name="output_proj")
+
+        next_contexts = loop_state.contexts.write(step, context)
+        next_head_weights = [loop_state.head_weights[i].write(step,
+                                                              head_weights[i])
+                             for i in range(self.n_heads)]
+
+        next_loop_state = MultiHeadLoopStateTA(
+            contexts=next_contexts,
+            head_weights=next_head_weights)
+
+        return context, next_loop_state
+
+    def attention_single_head(self, query: tf.Tensor,
+                              keys: tf.Tensor,
+                              values: tf.Tensor) -> Tuple[tf.Tensor,
+                                                          tf.Tensor]:
+        # shape: batch, time (similarities of attention keys in batch and time
+        # to the queries in the batch)
+        dot_product = tf.reduce_sum(
+            tf.expand_dims(query, 1) * keys, [-1])
+        energies = dot_product * self._scaling_factor
+
+        weights = tf.nn.softmax(energies)
+
+        if self.attention_mask is not None:
```

Comment:
Nemel by se ten softmax aplikovat az po maskovani? (aspon tak to uvadeji v tom paperu)

Reply:
One se to renormalizuje, takže je to skoro stejné jako kdyby bylo před až na nějakou zaokrouhlovaíc chybu. To je to, co se tady řešilo s tím `tf.where`.

Reply:
Tim, že ty logity můžou bejt záporný, tak je jedna cesta to napřed zesoftmaxit a pak renormalizovat.

Druhá cesta by byla vymaskovat to mínus nekonečnama:

```python
if self.attention_mask is not None:
  w_exp = tf.exp(weights)
  w_masked = weigths * self.attention_mask
  weights = tf.log(w_masked)
```

třetí cesta je použít `tf.where` ale teď nevim proč by.

Reply:
EDIT první verze kodu byla blbě, teď už je to snad dobre

Reply:
Ok, v pohode.

Reply:
tim editem sem myslel ten kod v minulym komentari (zacinajici slovy "Tim, že logity") A myslim si, že jestli je to dobre, tak bychom to meli pouzit - zbavi nas to toho `1e-08` stejně jako by nás toho zbavilo `tf.where` a navíc to bude bez potenciálního nebezpečí při použití ve while cyklu

Reply:
Možná by bylo ještě lepší udělat pomocnou funkci `masked_softmax`, abychom nemuseli ty exponenciely počítat dvakrát za sebou:

```python
def masked_softmax(logits, mask, dim):
  masked = tf.exp(logits) * mask
  return masked / tf.reduce_sum(masked, dim)
```